Mathlib/Order/RelSeries.lean

```diff
@@ -0,0 +1,198 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import Mathlib.Logic.Equiv.Fin
+import Mathlib.Data.List.Indexes
+import Mathlib.Data.Rel
+
+/-!
+# Series of a relation
+
+If `r` is a relation on `α` then a relation series of length `n` is a series
+`a_0, a_1, ..., a_n` such that `r a_i a_{i+1}` for all `i < n`
+
+-/
+
+variable {α : Type _} (r : Rel α α)
+
+/--
+Let `r` be a relation on `α`, a relation series of `r` of length `n` is a series
+`a_0, a_1, ..., a_n` such that `r a_i a_{i+1}` for all `i < n`
+-/
+structure RelSeries where
+/-- the number of inequalities in the series -/
+length : ℕ
+/-- the underlying function of a relation series -/
+toFun : Fin (length + 1) → α
+/-- adjacent elements are related by the said relation -/
+step : ∀ (i : Fin length), r (toFun <| Fin.castSucc i) <| toFun <| i.succ
+
+namespace RelSeries
+
+instance : CoeFun (RelSeries r) (fun x ↦ Fin (x.length + 1) → α) :=
+{ coe := RelSeries.toFun }
+
+instance : Preorder (RelSeries r) :=
+  Preorder.lift fun x => x.length
```

Comment:
That looks like a very poor notion of "less than". I would have rather expected `a ≤ b ↔ a.toList.sublist b.toList`. Do you have a reason behind this?

Reply:
This was originally for krull dimensions of a ring. So only the length of a chain is important.

Reply:
When considering `\covby` relation of submodules, this weak preorder will give the length of a module

Reply:
How will it "give" it?

Reply:
With this preorder we can also define type class of finite dimensional rings to be order with a top; infinite dimensional rings to be nonempty and without top.

Reply:
That sounds unnatural. I would expect you to use `no_max_order` instead of `no_top_order` (for the second one).